sdk/core/azure-core/src/main/java/com/azure/core/util/ConfigurationPropertyBuilder.java

```diff
@@ -0,0 +1,133 @@
+    private String[] environmentAliases = EMPTY_LIST;
+    private T defaultValue;
+    private boolean shared;
+    private boolean canLogValue;
+    private boolean required;
+
+    /**
+     * Constructs {@code ConfigurationPropertyBuilder} instance.
+     *
+     * <!-- src_embed com.azure.core.util.ConfigurationPropertyBuilder -->
+     * <pre>
+     * ConfigurationProperty&lt;SampleEnumProperty&gt; modeProperty =
+     *     new ConfigurationPropertyBuilder&lt;&gt;&#40;&quot;mode&quot;, SampleEnumProperty::fromString&#41;
+     *         .canLogValue&#40;true&#41;
+     *         .defaultValue&#40;SampleEnumProperty.MODE_1&#41;
+     *         .build&#40;&#41;;
+     * System.out.println&#40;configuration.get&#40;modeProperty&#41;&#41;;
+     * </pre>
+     * <!-- end com.azure.core.util.ConfigurationPropertyBuilder -->
+     *
+     * @param name name of the property.
+     * @param converter Converter used to map the configuration to {@code T}.
+     */
+    public ConfigurationPropertyBuilder(String name, Function<String, T> converter) {
+        this.name = Objects.requireNonNull(name, "'name' cannot be null");
+        this.converter = Objects.requireNonNull(converter, "'converter' cannot be null");
+    }
+
+    /**
+     * Sets default property value. {@code null} by default.
+     *
+     * @param defaultValue value to be returned by {@link Configuration#get(ConfigurationProperty)} if the property isn't found.
+     * @return the updated ConfigurationPropertyBuilder object.
+     */
+    public ConfigurationPropertyBuilder<T> defaultValue(T defaultValue) {
+        this.defaultValue = defaultValue;
+        return this;
+    }
+
+    /**
+     * Sets flag indicating that property can be provided in the shared configuration section
+     * in addition to client-specific configuration section.
+     * Default is {@code false}, indicating that property can only be provided in local configuration.
+     *
+     * @param shared If set to {@code true}, {@link Configuration#get(ConfigurationProperty)} will attempt to retrieve property from
+     *               local configuration and fall back to shared section, when local property is missing.
+     *               Otherwise, only local configuration will be checked.
+     * @return the updated ConfigurationPropertyBuilder object.
+     */
+    public ConfigurationPropertyBuilder<T> shared(boolean shared) {
+        this.shared = shared;
+        return this;
+    }
+
+    /**
+     * Sets flag indicating if property value can be logged.
+     * Default is {@code false}, indicating that property value will be redacted in logs.
+     *
+     * @param canLogValue If set to {@code true}, {@link Configuration#get(ConfigurationProperty)} will log property value,
+     *                    Otherwise, value is redacted.
+     * @return the updated ConfigurationPropertyBuilder object.
+     */
+    public ConfigurationPropertyBuilder<T> canLogValue(boolean canLogValue) {
```

Comment:
for future. should we consider ability to provide a sanitizer for property ? in case there's a value of logging it but some part must be erased.

Reply:
Good point, I can imagine it'd be useful with e.g. connections strings to help identify formatting issues. Assuming we do this, we'll need a different public method, e.g. ` public ConfigurationPropertyBuilder<T> valueSanitizer(Function<String, String>)` . 
I've experimented with it and it'd still be useful to keep `canLogValue` for simplicity (e.g. password in `ProxyOptions` or raw access keys)